lib/elixir/lib/code.ex

```diff
@@ -206,6 +206,237 @@ defmodule Code do
+
+  ## Return values
+
+    * `{:alias, charlist}` - the autocompletion is an alias, potentially
+      a nested one, such as `Hello.Wor` or `HelloWor`
+
+    * `{:alias_or_dot, charlist}` - the autocompletion is an alias or a dot
+     call, such as `Hello.` or `Hello.World.`
+
+    * `{:dot, inside_dot, charlist}` - the autocompletion is a dot
+      where `inside_dot` is either a `{:var, charlist}`, `{:alias, charlist}`,
+      `{:unquoted_atom, charlist}` or a `dot` itself. If a var is given,
+      this may either be a remote call or a map field access. Examples are
+      `Hello.wor`, `:hello.wor`, `hello.wor`, `Hello.nested.wor`, and
+      `hello.nested.wor`
+
+    * `{:dot_arity, inside_dot, charlist}` - the autocompletion is a dot arity
+      where `inside_dot` is either a `{:var, charlist}`, `{:alias, charlist}`,
+      `{:unquoted_atom, charlist}` or a `dot` itself. If a var is given,
+      it must be a remote arity. Examples are `Hello.world/`, `:hello.world/`,
+      and `hello.world/2`
+
+    * `{:dot_call, inside_dot, charlist}` - the autocompletion is a dot
+      call. This means parentheses or space have been added after the expression.
+      where `inside_dot` is either a `{:var, charlist}`, `{:alias, charlist}`,
+      `{:unquoted_atom, charlist}` or a `dot` itself. If a var is given,
+      it must be a remote call. Examples are `Hello.world(`, `:hello.world(`,
+      `Hello.world `, `hello.world(`, and `hello.world `
+
+    * `:expr` - may be any expression. Autocompletion may suggest an alias,
+      local or var
+
+    * `{:local_or_var, charlist}` - the autocompletion is a variable or a local
+      (import or local) call, such as `hello_wor`
+
+    * `{:local_arity, charlist}` - the autocompletion is a local (import or local)
+      call, such as `hello_world/`
+
+    * `{:local_call, charlist}` - the autocompletion is a local (import or local)
+      call, such as `hello_world(` and `hello_world `
+
+    * `:none` - no autocompletion possible
+
+    * `:unquoted_atom` - the autocompletion is an unquoted atom. This can be either
+      previous atoms or all available `:erlang` modules
+
+  ## Limitations
+
+    * There is no autocompletion of operators
+    * The current algorithm only consider the last line of the input
+    * Autocompletion does not track strings, sigils, etc
+    * Arguments of functions calls are not currently recognized
+
+  """
+  @spec autocomplete(List.Chars.t(), keyword()) ::
+          {:alias, charlist}
+          | {:alias_or_dot, charlist}
+          | {:dot, inside_dot, charlist}
+          | {:dot_arity, inside_dot, charlist}
+          | {:dot_call, inside_dot, charlist}
+          | {:local_or_var, charlist}
+          | {:local_arity, charlist}
+          | {:local_call, charlist}
+          | :none
+          | {:unquoted_atom, charlist}
+        when inside_dot: {:var, charlist} | {:alias, charlist} | {:unquoted_atom, charlist} | {:dot, inside_dot, charlist}
```

Comment:
What's the reason for working with charlists over binaries?

Reply:
Most of the parser code works with charlists because it is simpler to work with codepoints. Given the result of this will likely be piped into the parser, such as Code.string_to_quoted, I chose to return a charlist to avoid double or triple conversion.